---
title: PopHumanVar
output: html_document
runtime: shiny
---

[comment]: <> ( PopHumanVar is licensed under the GNU General Public License (GPL) v3.0 (https://github.com/ainacolovila/PopHumanVar/blob/master/LICENSE))
[comment]: <> (You can adjust the output dimensions by changing "options = list(width = "120%", height = 2200)" at the end of this markdown )


```{r global, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE)


# Libraries
# listPackages <- c("data.table", "dplyr", "plotly", "shiny", "shinyWidgets", "shinydashboard", "shinydashboardPlus", "shinyBS", "shinyjs", "shinyalert", "shinythemes", "shinyhelper", "shinycssloaders", "shinycustomloader", "splitstackshape", "DT", "stringr", "R.utils", "rehh", "vcfR", "intervals")
# To install only missing packages
# for(pp in listPackages){
#   if(!require(pp)) install.packages(pp)
# }

library(data.table)
library(dplyr)
library(plotly)
library(shiny)
library(shinyWidgets)
library(shinydashboard)
library(shinydashboardPlus)
library(shinyBS)
library(shinyjs)
library(shinyalert)
library(shinythemes)
library(shinyhelper)
library(shinycssloaders)
library(shinycustomloader)
library(splitstackshape)
library(DT)
library(stringr)
library(R.utils)
library(rehh)
library(vcfR)
library(intervals)


# Upload data from the folder
setwd(getwd())
AlldataSet <- fread("info.txt.gz", header = T)

# Modify dataset 
# Create column
AlldataSet$NumGWASStudies<-0
AlldataSet$NumDGNAsso <- 0

# Generate a column count
if(sum(!is.na(AlldataSet$accession))>0){
  AlldataSet$NumGWASStudies[!is.na(AlldataSet$accession)] <- do.call(rbind,lapply(
    strsplit(AlldataSet$accession[!is.na(AlldataSet$accession)],";"), function(x) length(x)))}
if(sum(!is.na(AlldataSet$diseaseId))>0){
	AlldataSet$NumDGNAsso[!is.na(AlldataSet$diseaseId)] <- do.call(rbind, lapply(
		strsplit(AlldataSet$diseaseId[!is.na(AlldataSet$diseaseId)],";"), function(x) length(x)))
}

# Get stedy variables from the data 
chrN <- unique(AlldataSet$chr)
minPos <- min(AlldataSet$physicalPos, na.rm = T)
maxPos <- max(AlldataSet$physicalPos, na.rm = T)
maxNofPmids <- max(AlldataSet$NofPmids, na.rm = T)
maxAgeMode <- max(c(AlldataSet$AgeMean_Jnt,AlldataSet$AgeMean_Mut,AlldataSet$AgeMean_Rec), na.rm = T)
maxAssoCounts <-  max(AlldataSet$NumGWASStudies, na.rm = T)

# FUNCTIONS
# Function to split collapsed dataframe
SplitDf <- function(dfFiltered, c1=c("gene_effect","impact","effect"),c2=c("effect"),s1=",",s2="&"){
	
	z1 = cSplit(dfFiltered, c1, sep=',',direction="long", makeEqual=F)

	if(!is.null(c2) & !is.null(s2)){
		z2 = cSplit(z1,c2,sep='&',direction="long")

		ind = which(names(z2) %in% c1,T)

		z2[, (ind) := lapply(.SD, as.character), .SDcols = ind]

		return(z2)

	}else{
		return(z1)  
	}
} 

# Funció colorByQuality
colorByQuality <- function(min, max){
	colorScale <- c("saddlebrown","darkred","red","chocolate1","orange",
				 "goldenrod1", "yellow2","lightgreen",
				 "green","forestgreen")
	calidade <- colorRampPalette(colorScale[min:max])
	return(calidade(10))
}

# Generate info boxes
getInfoBoxData <- function(stat,dfVariants){        
	# stat options: c('isafe','ihs','nsl','gwas','dgn', snpeff, clinvar,rgd)
	# Prioretize stats
	keyStats<- data.frame(
		'stat'=c('isafe','ihs','nsl','gwas','snpeff', 'rgd', 'clinvar','dgn'),
		'colNames'=c("isafe","ihsABS","nslABS","NumGWASStudies","rankSnpEFF","RDBranking","ClinVarRank","NofPmids"),'realColNames'=c("isafe","ihs","nsl","NumGWASStudies","effect","RDBranking","ClinicalSignificance","NofPmids"),'order'=c(-1L,-1L,-1L,-1L,1L,1L,1L,-1L))      
	
	orderStats   = append(as.vector(keyStats$colNames[keyStats$stat==stat]),as.vector(keyStats$colNames[keyStats$stat!=stat]))
	direction    = as.vector(keyStats$order[match(orderStats,keyStats$colNames)])
	finalColName = as.vector(keyStats$realColNames[keyStats$stat==stat])

	# Sort data according to stat
	setorderv(dfVariants, orderStats, order = direction,na.last=TRUE)

	# Search max/min value
  topValue  = dfVariants[[orderStats[1]]][1]
  topValues = dfVariants[dfVariants[[orderStats[1]]]==topValue,]
	if(topValue!=0 & !is.na(topValue) & nrow(topValues)>0){
		if(stat=='rgd' & nrow(topValues)>1){
			return(list(topValues$rsid[1],topValues[[finalColName]][1], nrow(topValues),
				paste0(as.vector(topValues$rsid),collapse="%0D%0A")))
		} else {
			return(list(topValues$rsid[1],topValues[[finalColName]][1], nrow(topValues)))
		}
	}
	else{return(NULL)}
}

# VARIABLES

#SNPEFF 
VEPinfo <- data.table(
  "inLabels"= c("splice_acceptor_variant", "splice_donor_variant", "stop_gained", "frameshift_variant",
			  "stop_lost", "start_lost", "bidirectional_gene_fusion", "disruptive_inframe_insertion",
			  "conservative_inframe_insertion", "disruptive_inframe_deletion", "conservative_inframe_deletion",
			  "missense_variant", "splice_region_variant", "start_retained_variant", "stop_retained_variant",
			  "synonymous_variant", "5_prime_UTR_premature_start_codon_gain_variant", "initiator_codon_variant",
			  "5_prime_UTR_variant", "3_prime_UTR_variant", "intron_variant", "non_coding_transcript_exon_variant",
			  "non_coding_transcript_variant", "upstream_gene_variant", "downstream_gene_variant",
			  "intragenic_variant", "intergenic_region"),
  "outLabels"=c("Splice acceptor variant", "Splice donor variant", "Stop gained", "Frameshift variant", "Stop lost",
			  "Start lost", "Bidirectional gene fusion", "Disruptive inframe insertion", "Conservative inframe insertion",
			  "Disruptive inframe deletion", "Conservative inframe deletion", "Missense variant", "Splice region variant",
			  "Start retained variant", "Stop retained variant", "Synonymous variant", "5-UTR premature start codon gain variant",
			  "Initiator codon variant", "5-UTR variant", "3-UTR variant", "Intron variant", "Non-coding transcript exon variant",
			  "Non-coding transcript variant", "Upstream gene variant ", "Downstream gene variant", "Intragenic variant",
			  "Intergenic region"), 'rank'=seq(1,27))

# ClinVar
ClinVarLabels <- data.frame(
	"ClinicalSign"= c('Pathogenic', 'Pathogenic/Likely pathogenic', 'Likely pathogenic',
	'risk factor','drug response, risk factor','drug response', 'affects','association','protective','Likely benign', 
	'Benign/Likely benign','Benign','other',"Uncertain significance","Conflicting interpretations of pathogenicity",
	"Conflicting interpretations from submitters","no interpretation for the single variant",'association not found',"not provided"), 
		'colors'= c('#D62728','#EF553B', '#E45756','#AB63FA','#625377','#316395','#3B7887', '#3B8775', 
						'#298538','#B6E880','#00CC96','#20701D','#A5C2B2',"#938D8A","#FFA15A", 
						"#FFA15A",'#938D8A','#938D8A','#938D8A'))

#RegulomeDB
regulomeLabels <- c('1a', '1b', '1c', '1d', '1e', '1f', '2a',
				'2b', '2c', '3a', '3b', '4','5','6','7')

ReguColors <- c("#D73027", "#F46D43", "#FDAE61", "#FEE090", "#E0F3F8", 
	"#ABD9E9", "#74ADD1", "#4575B4", "#B3E2CD", "#FDCDAC", "#CBD5E8", 
	"#F4CAE4", "#E6F5C9", "#FFF2AE")

# Clocks-GEVA
GEVAclocks <- data.frame(
	"modelClocks" = c("Jnt","Mut","Rec"),
	"plotTitle" = c("Joint clock", "Mutation clock", "Recombination clock"))
```


```{r app, echo=FALSE}
shinyApp(
  ui = dashboardPage(
  dashboardHeader(disable=TRUE),
  dashboardSidebar(
    sidebarMenu(
    	menuItem("Coordinates",
    		    icon = icon("search"),
    		    tabName = "dashboard",
    		    br(),
    		    "  GRCh37/hg19",
    		    br(),
    		    textInput('minPosInput', 'Start Position', 
    		    		value = minPos),
    		    textInput('maxPosInput', 'End Position', 
    		    		value = maxPos),
    		    # Input per fer zoom out
    		    tags$div(align = 'right',
    		    	    radioGroupButtons(
    		    	    	inputId = 'zoomCoordenates',
    		    	    	choices =c(`<i class="fa fa-search-minus fa-2x"  aria-hidden="true"></i>`= "-1",
    		    	    			 `<i class="fa fa-search-plus fa-2x"  aria-hidden="true"></i>`= "1"),
    		    	    	selected = '',
    		    	    	size='xs',
    		    	    	direction = "horizontal",
    		    	    	individual=T, disabled = F))
			),
			br(),
			menuItem("Selection (iHS & nSL)",
				icon = icon("stream"), 
				tabName = "selection"),
			menuItem("Favored Mutation (iSAFE)",
				icon = icon("asterisk"), 
				tabName = "isafe"),
			menuItem("Functional Description",
				icon = icon("layer-group"), 
				tabName = "functional"),
			menuItem("Age Information",
				icon = icon("hourglass"), 
				tabName = "age"),
			menuItem("Summary Report",
				icon = icon("file-alt"), 
				tabName = "report"),
			menuItem("Raw Table",
				icon = icon("table"), 
				tabName = "table"),
			menuItem("Script summary",
				icon = icon("code"), 
				tabName = "code")
	  	)
  ),
  dashboardBody(
     tabItems(
       tabItem(tabName = "selection",
               fluidPage(
                 fluidRow(align="center",HTML('<h2><i class="fas fa-stream"></i> <b>Selection (iHS & nSL)</b></h2>')),
                 fluidRow(align="left",
                          HTML('<h3><b>iHS distribution</b></h3>'),
                          br(),
                          box(width=12,
                          title=HTML('<h4><b>Filters iHS</b></h4>'),
                          column(6,
                                 br(),
                                 awesomeCheckbox(inputId = 'extremeValuesInput_ihs',
                                                 label="Only extreme iHS Values",
                                                 value = FALSE, status="primary")),
                          column(6,
                                 numericInput("pvalue_ihs",
                                              "Top extreme %:", 
                                              value =0.005, min = 0.0001, max =0.05, step=0.0001)
                                 )),
                          br(),
                           box(width=12,
                            title=HTML('<h4><b>Plot iHS</b></h4>'),
                            plotlyOutput("ihsPlot")),
                          br()
                          ),
                 fluidRow(align="left",
                          HTML('<h3><b>nSL distribution</b></h3>'),
                          br(),
                          box(width=12,
                          title=HTML('<h4><b>Filters nSL</b></h4>'),
                          column(6,
                                 br(),
                                 awesomeCheckbox(inputId = 'extremeValuesInput_nsl',
                                                 label="Only extreme iHS Values",
                                                 value = FALSE, status="primary")),
                          column(6,
                                 numericInput("pvalue_nsl",
                                              "Top extreme %:", 
                                              value =0.005, min = 0.0001, max =0.05, step=0.0001)
                                 )),
                          br(),
                           box(width=12,
                            title=HTML('<h4><b>Plot nSL</b></h4>'),
                            plotlyOutput("nslPlot")),
                          br()
                          )
                 )
               ),
       tabItem(tabName = "isafe",
               fluidPage(
                 fluidRow(align="center",HTML('<h2><i class="fas fa-asterisk"></i> <b>Favored Mutation (iSAFE)</b></h2>')),
                 fluidRow(align="left",
                          HTML('<h3><b>iSAFE distribution</b></h3>'),
                          br(),
                          box(width=12,
                              title=HTML('<h4><b>Filters iSAFE</b></h4>'),
                          column(6,
                                 br(),
                                 awesomeCheckbox(inputId = 'extremeValuesInput_isafe',
                                                 label="Only extreme iHS Values",
                                                 value = FALSE, status="primary")),
                          column(6,
                                 numericInput("pvalue_isafe",
                                              "Top extreme %:", 
                                              value =0.005, min = 0.0001, max =0.05, step=0.0001)
                                 )),
                          br(),
                           box(width=12,
                            title=HTML('<h4><b>Plot iSAFE</b></h4>'),
                            plotlyOutput("isafePlot")),
                          br()
                          )
                 )
               ),
       tabItem(tabName = "functional",
               fluidPage(
                 fluidRow(align="center",HTML('<h2><i class="fas fa-layer-group"></i> <b>Functional Description</b></h2>')),
                 fluidRow(align="left",
                          HTML('<h3><b>Variant Effect (SnpEff)</b></h3>'),
                          br(),
                          box(width=12,
                              title= HTML('<h4><b>Filters SnpEff</b></h4>'),
                              collapsible = TRUE,
                              collapsed=TRUE,
                              actionButton('selectAllVEPInput', 'Select All', width='25%'),
                              br(),br(),
                              awesomeCheckboxGroup(inputId = 'VEPInput',
                                                   'SnpEff Consequences',
                                                   choices = as.vector(VEPinfo$outLabels),
                                                   selected = as.vector(VEPinfo$outLabels[VEPinfo$rank<12]),
                                                   status='primary',
                                                   inline = FALSE)
                              ),
                          br(),
                          box(width=12,
                              collapsible = TRUE,
                              collapsed=FALSE,
                              title=HTML('<h4><b>Plot SnpEff</b></h4>'),
                              withSpinner(
                              	plotlyOutput("snpeffPlot"),
                              	type = 4,
                              	color = "#d33724", 
                              	size = 0.7)
                              ),
                          br()
                          ),
                 fluidRow(align="left",
                          HTML('<h3><b>Regulome DB</b></h3>'),
                          br(),
                          box(width=12,
                              title= HTML('<h4><b>Filters RegulomeDB</b></h4>'),
                              collapsible = TRUE,
                              collapsed=TRUE,
                              column(3),
                              column(6,
                                     sliderTextInput('RegulomeInput', 
                                                     h5('Regulome Score:'),
                                                     choices = regulomeLabels, 
                                                     selected = c('1a','7'),
                                                     grid = TRUE)),
                              column(3)),
                          br(),
                          box(width=12,
                              collapsible = TRUE,
                              collapsed=TRUE,
                              title=HTML('<h4><b>Plot Regulome DB</b></h4>'),
                              withSpinner(
                              	plotlyOutput("reguDBPlot"),
                              	type = 4,
                              	color = "#d33724", 
                              	size = 0.7)
                              ),
                          br()
                          ),
                 fluidRow(align="left",
                          HTML('<h3><b>ClinVar</b></h3>'),
                          br(),
                           box(width=12,
                              title= HTML('<h4><b>Filters ClinVar</b></h4>'),
                              collapsible = TRUE,
                              collapsed=TRUE,
                              actionButton('selectAllClinVarInput', 'Select All', width='25%'),br(),br(),
                              awesomeCheckboxGroup(inputId = 'ClinVarInput',
                                                   'Clinical Significances:',
                                                   choices = as.vector(ClinVarLabels$ClinicalSign),
                                                   selected = as.vector(ClinVarLabels$ClinicalSign),
                                                   status='primary', inline = FALSE)
                              ),
                          br(),
                          box(width=12,
                              collapsible = TRUE,
                              collapsed=TRUE,
                              title=HTML('<h4><b>Plot ClinVar</b></h4>'),
                              withSpinner(
                              	plotlyOutput("clinvarPlot"),
                              	type = 4,
                              	color = "#d33724", 
                              	size = 0.7)
                              ),
                          box(width=12,
                              collapsible = TRUE,
                              collapsed=TRUE,
                              title=HTML('<h4><b>ClinVar Pie</b></h4>'),
                              withSpinner(
                              	plotlyOutput("clinvarPie",width = "90%", height = "500px"),
                              	type = 4,
                              	color = "#d33724", 
                              	size = 0.7)
                              ),
                          br()
                          ),
                 fluidRow(align="left",
                          HTML('<h3><b>GWAS Catalog</b></h3>'),
                         br(),
                          box(width=12,
                              title= HTML('<h4><b>Filters GWAS Catalog</b></h4>'),
                              collapsible = TRUE,
                              collapsed=TRUE,
                              column(3),
                              column(6,
                              sliderInput('GWASInput',
                                          h5('Filter by Association Counts:'),
                                          min = 0,
                                          max = maxAssoCounts,
                                          value = c(0,maxAssoCounts),
                                          step = 1)
                              ),
                              column(3)
                              ),
                          br(),
                          box(width=12,
                              collapsible = TRUE,
                              collapsed=TRUE,
                              title=HTML('<h4><b>Plot GWAS Catalog</b></h4>'),
                              withSpinner(
                              	plotlyOutput("gwasPlot"),
                              	type = 4,
                              	color = "#d33724", 
                              	size = 0.7)
                              ),
                          box(width=12,
                              collapsible = TRUE,
                              collapsed=TRUE,
                              title=HTML('<h4><b>GWAS Catalog Pie</b></h4>'),
                               withSpinner(
                              	plotlyOutput("gwasPie",width = "90%", height = "500px"),
                              	type = 4,
                              	color = "#d33724", 
                              	size = 0.7)
                              ),
                          br()
                          ),
                 fluidRow(align="left",
                          HTML('<h3><b>DisGeNET</b></h3>'),
                         br(),
                         box(width=12,
                              title= HTML('<h4><b>Filters DisGeNET</b></h4>'),
                              collapsible = TRUE,
                              collapsed=TRUE,
                              column(4,
                                       sliderInput('DSIInput',
                                                   h5('Disease Specificity Index:'),
                                                   min = 0, max = 1, step=0.05, value = c(0, 1))),
                                column(4,
                                       sliderInput('EInput',
                                                   h5('Evidence Index:'),
                                                   min = 0, max = 1, step=0.05, value = c(0, 1))),
                                column(4,
                                       sliderInput('NofPmidsInput',
                                                   h5('Number of Pubmed IDs:'),
                                                   min = 0, max = max(maxNofPmids, na.rm = T),
                                                   value = c(0,max(maxNofPmids, na.rm = T)))
                                       )
                              ),
                          br(),
                          box(width=12,
                              collapsible = TRUE,
                              collapsed=TRUE,
                              title=HTML('<h4><b>Plot DisGeNET</b></h4>'),
                              withSpinner(
                              	plotlyOutput("dgnPlot"),
                              	type = 4,
                              	color = "#d33724", 
                              	size = 0.7)
                              ),
                          br()
                          )
                 )
               ),
       tabItem(tabName = "age",
               fluidPage(
                 fluidRow(align="center",HTML('<h2><i class="fas fa-hourglass"></i> <b>Age Information</b></h2>')),
                 fluidRow(align="left",
                          HTML('<h3><b>Atlas of Variant Age</b></h3>'),
                          br(),
                          box(width=12,
                              title= HTML('<h4><b>Filters AVA</b></h4>'),
                              collapsible = TRUE,
                              collapsed=FALSE,
                              column(6,
                                     br(),
                                     pickerInput(
                                       'AgeModelInput',
                                       'Clock Model:',
                                       choices = c("Mutation clock"='Mut',
                                                   "Recombination clock"='Rec',
                                                   "Joint clock"='Jnt'),
                                       selected = 'Jnt'
                                       ),
                                     numericInput("GenYears",
                                                  "Years per generation",
                                                  value =0, min = 0, max = 50, step=1),
                                     awesomeCheckbox(inputId = 'errorBars',
                                                     label="Show error bars",
                                                     value = FALSE,
                                                     status="primary")
                                     ),
                              column(6,
                                     sliderInput(
                                       'AgeModeInput',
                                       'Age Mode (generations ago):',
                                       min = 0,
                                       max = maxAgeMode,
                                       value = c(0,maxAgeMode/4)),
                                     sliderInput(
                                       'qualScoreInput',
                                       h5('Quality Score:'),
                                       min = 0,
                                       max = 1,
                                       value = c(0.85, 1),
                                       step=0.05
                                       )
                                     )
                              ),
                          br(),
                          box(width=12, 
                              title=HTML('<h4><b>Plot Atlas of Variant Age</b></h4>'),
                              withSpinner(
                              	plotlyOutput("agePlot", height = "700px"),
                              	type = 4,
                              	color = "#d33724", 
                              	size = 0.7)
                              ),
                          br()
                          ))),
       tabItem(tabName = "report",
               fluidPage(
                 fluidRow(align="center",HTML(
                 	paste0('<h2><i class="fas fa-file-alt"></i> <b>Summary Report</b></h2><br>',
                 		   htmlOutput("DescriptionRegionTitle")))),
                 fluidRow(
                 	align = "center",
                 	column(1),
                 	column(10,
                 		  # add buttons to further explore the region
                 		  div(style="display:inline-block;width:49%;text-align: right;",uiOutput("pophumanscanLink")),
                 		  div(style="display:inline-block;width:49%;text-align: right;",uiOutput("pophumanLink"))),
                 		column(1)
					),
                 fluidRow(align="left",
                 	    box(width=12,
                              collapsible = TRUE,
                              collapsed=TRUE,
                              title= HTML('<h4><b>Filters Summary</b></h4>'),
                              column(5,
                                     numericInput("Overview_pvalue_ihs",
                                              h5("iHS Top extreme %:"), 
                                              value =0.005, min = 0.0001, max =0.05, step=0.0001),
                              	  br(),
                                     numericInput("Overview_pvalue_nsl",
                                              h5("nSL Top extreme %:"), 
                                              value =0.005, min = 0.0001, max =0.05, step=0.0001),
                              	  br(),
                                     numericInput("Overview_pvalue_isafe",
                                              h5("iSAFE Top extreme %:"), 
                                              value =0.005, min = 0.0001, max =0.05, step=0.0001),
                              	  br()
                                 ),
                              column(1),
                              column(6,
                              	  pickerInput(
									'Overview_Clock',
									h5('Clock Model:'),
									choices = c("Mutation clock"='Mut', 
										"Recombination clock"='Rec', 
										"Joint clock"='Jnt'),
									selected = 'Jnt'
								),br(),
                                     sliderInput(
                                       'Overview_AgeMode',
                                       h5('Age Mode (generations ago):'),
                                       min = 0,
                                       max = maxAgeMode,
                                       value = c(0,maxAgeMode/4)),
                                     sliderInput(
                                       'Overview_qualScore',
                                       h5('Quality Score:'),
                                       min = 0,
                                       max = 1,
                                       value = c(0.85, 1),
                                       step=0.05)
                                     )
                              ),
					br(),
					box(width=12,
                             collapsible = TRUE,
                             collapsed=FALSE,
                             title=HTML('<h4><b>Plot Summary</b></h4>'),
					    withSpinner(
                              	plotlyOutput("reportPlot", height = "700px"),
                              	type = 4,
                              	color = "#d33724", 
                              	size = 0.7)
                              ),
					br(),
					box(width=12,
                             collapsible = TRUE,
                             collapsed=FALSE,
                             title=HTML('<h4><b>Summary</b></h4>'),
					    fluidRow(
					    	 infoBoxOutput("info_isafe",width = 8),
					    	 infoBoxOutput("info_iHS",width = 8), 
					    	 infoBoxOutput("info_nsl",width = 8), 
					    	 infoBoxOutput("info_snpeff",width = 8), 
					    	 infoBoxOutput("info_ReguDB",width = 8), 
					    	 infoBoxOutput("info_ClinVar",width = 8), 
					    	 infoBoxOutput("info_GWAScat",width = 8), 
					    	 infoBoxOutput("info_DGN",width = 8)
					    )
					),
					br(),
                         box(width=12,
                             collapsible = TRUE,
                             collapsed=FALSE,
                             title= HTML('<h3><b>Top 20 Variants</b></h3>'),
                             withSpinner(
                             	dataTableOutput('tablePrior'),
                             	type = 4,
                             	color = "#d33724", 
                             	size = 0.7)
                             ),
					br(),
					box(width=12,
                             collapsible = TRUE,
                             collapsed=TRUE,
                             title= HTML('<h3><b>Haplotype Visualization</b></h3>'),
					    column(12,
					    	  withSpinner(
								type = 4,
								color = "#d33724",
								size = 0.7,
								plotOutput("haplo1", height=200)
							),
							withSpinner(
								type = 4,
								color = "#d33724",
								size = 0.7,
								plotOutput("haplo3", height=350)
							),
							withSpinner(
								type = 4,
								color = "#d33724",
								size = 0.7,
								plotOutput("haplo2", height=475)
							)
					)
                             ),
					br()
					))),
       tabItem(tabName = "table",
               fluidPage(
                 fluidRow(align="center",HTML('<h2><i class="fas fa-table"></i> <b>Data Table</b></h2>')),
                 fluidRow(align="center",
                 	    withSpinner(
                              	dataTableOutput('table_all'),
                              	type = 4,
                              	color = "#d33724", 
                              	size = 0.7)
                          ),
                 br()
               )
               ),
       tabItem(tabName = "code",
               fluidPage(
                 fluidRow(align="center",HTML('<h2><i class="fas fa-code"></i> <b>Script</b></h2>')),
                 fluidRow(align="left",
                          box(width=12,
                              HTML(paste0('<pre>',
                                          includeHTML(list.files(pattern = '\\.sh')),
                                          '</pre>'))
                              ),
                          br()
                          )
                 )
               )
       )
       )
  ),

  server = function(input, output, session) {
    dataSet <- reactive({AlldataSet %>% dplyr::filter(physicalPos>=input$minPosInput,physicalPos<=input$maxPosInput)})
    
    # Zoom buttons
    observeEvent(input$zoomCoordenates,{
    	
		# Compute increment of coordinates input (100k change)
		newStart <- as.numeric(input$minPosInput)+(as.numeric(input$zoomCoordenates)*100000)
		newEnd <- as.numeric(input$maxPosInput)-(as.numeric(input$zoomCoordenates)*100000)

		# If chromosome end reached, keep last position with data
		if(newStart < minPos){newStart<-minPos}
		if(newEnd > maxPos){newEnd <-maxPos}
		
		# Check limits and change number
		if(newStart < newEnd){
				updateTextInput(session, "minPosInput", label = 'Start Position', value = newStart)
				updateTextInput(session, "maxPosInput", label = 'End Position', value = newEnd)
		} else {
			sendSweetAlert(
    				session = session,
    				title = "Warning!",
					closeOnClickOutside = FALSE,
					text = tags$span(
						tags$h2("You can not ZOOM IN anymore."),tags$br(),
						" Since the zoom is of 100kb, next increment would invert the ",
						tags$b("START"),"and",tags$b("END"), "coordinates.",tags$br(),
						"If you would like a smaller increment, please, update it manually."),
    				html = TRUE,
    				type = "warning"
 			)
		}
		
		# After action, unclick botton
		updateRadioGroupButtons(session, "zoomCoordenates", selected=character(0))
	})
    
    #█ Selection Plots
    output$ihsPlot <- renderPlotly({
      # Make data subset
      iHSsubset <- dataSet() %>% dplyr::select(physicalPos, rsid,standarizediHS) %>%
      dplyr::filter(!is.na(standarizediHS))
      iHSsubset$AD <- ifelse(iHSsubset$standarizediHS>0,"D","A")
    
      tresh <- unname(quantile(abs(iHSsubset$standarizediHS),1-input$pvalue_ihs))
      # Boxplot
      figBoxplot <- iHSsubset %>% dplyr::filter(abs(standarizediHS)>=tresh)%>%
      plot_ly(
        type = 'scatter',
  			mode = 'markers',
  			x = 1,
  			y = ~standarizediHS,
  			color="orange",
  			text = ~paste0('<b>',rsid,'</b><br>', physicalPos),
  			hovertemplate = paste(
  				"%{text}<br>",
  				"<b>iHS:</b>: %{y}<br>"),
  			showlegend = FALSE) %>% layout(
        	xaxis = list(
        		title = "",
        		zeroline = FALSE,
        		showline = FALSE,
        		showticklabels = FALSE,
        		showgrid = FALSE
        		)
        	)
  	  # if Filter by extreme, change boxplot colors
  	  if(input$extremeValuesInput_ihs==TRUE){
  	    figBoxplot <- figBoxplot %>% add_trace(
  	      data = iHSsubset,
  			  y = ~standarizediHS,
  			  type="box",
  			  boxpoints=FALSE,
  			  showlegend = FALSE,
  			  fillcolor ="#e7efef",
  			  line = list(color = "#909495"))
  	  }else{
  		  figBoxplot <- figBoxplot %>% add_trace(
  			  data = iHSsubset,
  			  y = ~standarizediHS,
  			  type="box",
  			  boxpoints=FALSE,
  			  showlegend = FALSE,
  			  fillcolor ="lightblue",
  			  line = list(color = "#909495")
  			  )
  	}
	
  # Distribution along positions
  figScatter <- iHSsubset %>% 
  	plot_ly(x = ~physicalPos, y = ~standarizediHS) %>%
  	add_trace(type = 'scatter',
              mode = 'markers',
              color = ~AD,
    		colors= c("firebrick","darkblue"),
              text = ~paste('<b>', rsid,'</b>:',
                            round(standarizediHS,2),
              		    '<br><b>', AD, '</b>'
                            ),
              hoverinfo = 'text',
              showlegend = F) %>%
    layout(
			showlegend = FALSE,
			shapes = list(
				list(
					type = "line", 
					x0 = 0, 
					x1 = 1, 
					xref = "paper",
					y0 = tresh, 
					y1 = tresh, 
					line = list(color = "orange"),
					width = 4, 
					dash = 'dot',
					opacity = 0.4
				),
				list(
					type = "line", 
					x0 = 0, 
					x1 = 1, 
					xref = "paper",
					y0 = -tresh, 
					y1 = -tresh, 
					line = list(color = "orange"),
					width = 4, 
					dash = 'dot',
					opacity = 0.4
				)),
			xaxis = list(title = 'Position (bp)', range = c(input$minPosInput, input$maxPosInput)),
			yaxis = list(title = "iHS"))
  
  	# if Filter by extreme, change boxplot colors
	if(input$extremeValuesInput_ihs==TRUE){
		figScatter <- figScatter %>% 
			add_trace(
				data=iHSsubset[abs(iHSsubset$standarizediHS)<=tresh],
				x = ~physicalPos, 
				y = ~standarizediHS,
				type = 'scatter',
					mode = 'markers',
					colors= "lightgrey",
					showlegend = F)
	}
  
  # Otput both plots:
  fig <- subplot(figBoxplot, figScatter, widths = c(0.2, 0.8)) %>% 
  layout(title = 'iHS Distribution')
  fig
  })
    output$nslPlot <- renderPlotly({
      # Make data subset
      nSLsubset <- dataSet() %>% dplyr::select(physicalPos, rsid, standarizednSL) %>%
        dplyr::filter(!is.na(standarizednSL))
      nSLsubset$AD <- ifelse(nSLsubset$standarizednSL>0,"D","A")
      
      tresh <- unname(quantile(abs(nSLsubset$standarizednSL),1-input$pvalue_nsl))
      
      # Boxplot
      figBoxplotnsl <- nSLsubset %>% dplyr::filter(abs(standarizednSL)>=tresh)%>%
        plot_ly(
          type = 'scatter',
    			mode = 'markers',
    			x = 1,
    			y = ~standarizednSL,
    			color="orange",
    			text = ~paste0('<b>',rsid,'</b><br>', physicalPos),
    			hovertemplate = paste(
    				"%{text}<br>",
    				"<b>nSL:</b>: %{y}<br>"),
    			showlegend = FALSE) %>% layout(
          	xaxis = list(
          		title = "",
          		zeroline = FALSE,
          		showline = FALSE,
          		showticklabels = FALSE,
          		showgrid = FALSE
          		)
          	)
    	  # if Filter by extreme, change boxplot colors
    	  if(input$extremeValuesInput_nsl==TRUE){
    	    figBoxplotnsl <- figBoxplotnsl %>% add_trace(
    	      data = nSLsubset,
    			  y = ~standarizednSL,
    			  type="box",
    			  boxpoints=FALSE,
    			  showlegend = FALSE,
    			  fillcolor ="#e7efef",
    			  line = list(color = "#909495"))
    	  }else{
    		  figBoxplotnsl <- figBoxplotnsl %>% add_trace(
    			  data = nSLsubset,
    			  y = ~standarizednSL,
    			  type="box",
    			  boxpoints=FALSE,
    			  showlegend = FALSE,
    			  fillcolor ="lightblue",
    			  line = list(color = "#909495")
    			  )
    	}
  	
       
      # Distribution along positions
      figScatternsl <- nSLsubset %>% plot_ly(x = ~physicalPos, y = ~standarizednSL) %>%
        add_trace(type = 'scatter',
                mode = 'markers',
                color = ~AD,
                colors= c("firebrick","darkblue"),
                text = ~paste('<b>', rsid,'</b>:',
                              round(standarizednSL,2),
                		    '<br><b>', AD, '</b>'
                              ),
                hoverinfo = 'text',
                showlegend = F) %>%
      layout(
  			showlegend = FALSE,
  			shapes = list(
  				list(
  					type = "line", 
  					x0 = 0, 
  					x1 = 1, 
  					xref = "paper",
  					y0 = tresh, 
  					y1 = tresh, 
  					line = list(color = "orange"),
  					width = 4, 
  					dash = 'dot',
  					opacity = 0.4
  				),
  				list(
  					type = "line", 
  					x0 = 0, 
  					x1 = 1, 
  					xref = "paper",
  					y0 = -tresh, 
  					y1 = -tresh, 
  					line = list(color = "orange"),
  					width = 4, 
  					dash = 'dot',
  					opacity = 0.4
  				)),
  			xaxis = list(title = 'Position (bp)', range = c(input$minPosInput, input$maxPosInput)),
  			yaxis = list(title = "nSL"))
    
    	# if Filter by extreme, change boxplot colors
  	if(input$extremeValuesInput_nsl==TRUE){
  		figScatternsl <- figScatternsl %>% 
  			add_trace(
  				data=nSLsubset[abs(nSLsubset$standarizednSL)<=tresh],
  				x = ~physicalPos, 
  				y = ~standarizednSL,
  				type = 'scatter',
  					mode = 'markers',
  					colors= "lightgrey",
  					showlegend = F)
  	}
    
      # Otput both plots:
      fignsl <- subplot(figBoxplotnsl, figScatternsl, widths = c(0.2, 0.8)) %>% 
      layout(title = 'nSL Distribution')
      fignsl
  })
    
    # Fav mutation Plots
    output$isafePlot <- renderPlotly({
      # Make data subset
      isafesubset <- dataSet() %>% dplyr::select(physicalPos, rsid, iSAFE) %>%
        dplyr::filter(!is.na(iSAFE), iSAFE!=0)
      tresh <- unname(quantile(isafesubset$iSAFE,1-input$pvalue_isafe))
      isafesubset$extreme <- ifelse(isafesubset$iSAFE>=tresh,"Extreme", "Not Extreme")
      
      # Boxplot
      figBoxplotisafe <- isafesubset %>% dplyr::filter(iSAFE>=tresh) %>% 
      	plot_ly(
          type = 'scatter',
          mode = 'markers',
          x=1,
          y = ~iSAFE,
          color="orange",
          text = ~paste0('<b>',rsid,'</b><br>', physicalPos),
          hovertemplate = paste(
          	"%{text}<br>",
          	"<b>iSAFE:</b>: %{y}<br>"),
          showlegend = FALSE) %>% layout(
          	xaxis = list(
          		title = "",
          		zeroline = FALSE,
          		showline = FALSE,
          		showticklabels = FALSE,
          		showgrid = FALSE
          		)
          	)
    	# if Filter by extreme, change boxplot colors
    	if(input$extremeValuesInput_isafe==TRUE){
    	  figBoxplotisafe <- figBoxplotisafe %>% 
    	  	add_trace(
    	  		data = isafesubset,
    			y = ~iSAFE,
    			type="box",
    			boxpoints=FALSE,
    			showlegend = FALSE,
    			fillcolor ="#e7efef",
    			line = list(color = "#909495"))
    	}else{
    		figBoxplotisafe <- figBoxplotisafe %>% add_trace(
    			data = isafesubset,
    			y = ~iSAFE,
    			type="box",
    			boxpoints=FALSE,
    			showlegend = FALSE,
    			fillcolor ="lightblue",
    			line = list(color = "#909495"))
    	}
    
      # Distribution along positions
      figScatterisafe <- isafesubset %>% plot_ly(x = ~physicalPos, y = ~iSAFE) 
      
      if(input$extremeValuesInput_isafe==TRUE){
        figScatterisafe <- figScatterisafe %>% 
          add_trace(type = 'scatter',
                  mode = 'markers',
                  color=~extreme,
                  colors=c("orange","#e7efef"),
                  text = ~paste('<b>', rsid,'</b>:',
                                round(iSAFE,2)),
                  hoverinfo = 'text',
                  showlegend = F) 
      }else{
        figScatterisafe <-figScatterisafe %>% add_trace(type = 'scatter',
                  mode = 'markers',
                  color='blue', 
                  text = ~paste('<b>', rsid,'</b>:',
                                round(iSAFE,2)),
                  hoverinfo = 'text',
                  showlegend = F) 
      }
      figScatterisafe <- figScatterisafe %>% 
        layout(
    			showlegend = FALSE,
    			shapes = list(
    				list(
    					type = "line",
    					x0 = 0,
    					x1 = 1,
    					xref = "paper",
    					y0 = tresh,
    					y1 = tresh,
    					line = list(color = "orange"),
    					width = 4,
    					dash = 'dot',
    					opacity = 0.4
    				)),
    			xaxis = list(title = 'Position (bp)', range = c(input$minPosInput, input$maxPosInput)),
    			yaxis = list(title = "iSAFE"))
    
      # Otput both plots:
      figisafe <- subplot(figBoxplotisafe, figScatterisafe, widths = c(0.2, 0.8)) %>%
      layout(title = 'iSAFE Distribution')
      figisafe
    })
    
    # Functional Plots
    # VEP
    observe({
		if (input$selectAllVEPInput > 0){
			if (input$selectAllVEPInput %% 2 == 0){
				updateCheckboxGroupInput(session, 'VEPInput', selected = '')
				updateActionButton(session, inputId="selectAllVEPInput", label = "Select All")
			}
			else {
				updateCheckboxGroupInput(session, 'VEPInput', selected = VEPinfo$outLabels)
				updateActionButton(session, inputId="selectAllVEPInput", label = "Deselect All")
			}
		}
	})
    output$snpeffPlot<- renderPlotly({
      
      # Make data subset
      vepsubset <- dataSet() %>% dplyr::select(physicalPos, rsid, effect, impact, gene_effect) %>%
        dplyr::filter_at(vars(effect, impact), any_vars(!is.na(.)))
      
      vepDcast <- SplitDf(vepsubset)
      vepDcast$rank <- match(vepDcast$effect, VEPinfo$inLabels)
      vepDcast$outLabs <- VEPinfo$outLabels[vepDcast$rank]

      vepDcast %>%
        # dplyr::filter(outLabs %in% input$VEPInput) %>% 
      	plot_ly(x = ~physicalPos, 
      	        y = ~outLabs,
      	        color= ~impact,
      	        type="scatter",
      	        mode="markers",
      	        text = ~paste0('<b>', rsid,'</b><br> ', outLabs, '<br>', impact),
      	        hoverinfo = 'text',
      	        marker = list(size = 8,
      	                      line = list(color = 'black', width = 1.5))) %>% 
        layout(title="",
               yaxis = list(
                 title = '',
                 type = 'category',
                 autorange="reversed",
                 categoryorder="array",
                 categoryarray=VEPinfo$outLabels),
               xaxis = list(title = 'Position (bp)',
               		   range = c(input$minPosInput, input$maxPosInput))
               )
    })
    
    output$reguDBPlot<- renderPlotly({
      filtered <- dataSet() %>% 
        dplyr::select(physicalPos, rsid, TFbinding, DNasePeak, motif, DNaseFootprint, 
                      eQTL, matchedTFmotif, matchedDNaseFootprint, ranking) %>% 
        dplyr::filter_at(vars(-c("physicalPos", "rsid")), any_vars(!is.na(.)))
      
      toPlot <- filtered[ranking >= input$RegulomeInput[1] & ranking <= input$RegulomeInput[2]]
      
      toPlot %>% plot_ly(x = ~physicalPos, 
                         y = ~ranking,
                         color = ~ranking,
                         colors=ReguColors,
                         name = 'rank', 
                         type = 'scatter', 
                         mode = 'markers',
                         text = ~paste0('<b>', rsid,'</b><br>', 
                                        physicalPos,'<br><br>',
                                        ifelse(TFbinding==1, "TFbinding <br>",""),
                                        ifelse(DNasePeak==1, "DNasePeak <br>",""),
                                        ifelse(motif==1, "motif <br>",""),
                                        ifelse(DNaseFootprint==1, "DNaseFootprint <br>",""),
                                        ifelse(eQTL==1, "eQTL <br>",""),
                                        ifelse(matchedTFmotif==1, "matchedTFmotif","")),
                         hoverinfo = 'text',
                         marker = list(size=7,line = list(color = '#404241', width = 1.5))) %>% 
        layout(yaxis = list(title = 'Score',
                            type = 'category',
                            autorange="reversed"),
               xaxis = list(title = 'Position (bp)',
				  		   range = c(input$minPosInput, input$maxPosInput)),
               showlegend = FALSE)
})
    
    	# Clinvar
    observe({
		if (input$selectAllClinVarInput > 0){
			if (input$selectAllClinVarInput %% 2 == 0){
				updateCheckboxGroupInput(session, 'ClinVarInput', selected = '')
				updateActionButton(session, inputId="selectAllClinVarInput", label = "Select All")
			}
			else {
				updateCheckboxGroupInput(session, 'ClinVarInput', selected = ClinVarLabels$ClinicalSign)
				updateActionButton(session, inputId="selectAllClinVarInput", label = "Deselect All")
			}
		}
	})
    output$clinvarPlot<- renderPlotly({
    	
    	# Filter data
	toPlot <- dataSet() %>% dplyr::select(physicalPos, rsid, ClinicalSignificance, ClinSigSimple, PhenotypeList) %>%
		dplyr::filter(ClinicalSignificance %in% input$ClinVarInput, !is.na(rsid)) %>%
		dplyr::filter_at(vars(-c("physicalPos", "rsid")), any_vars(!is.na(.)))
	
	toPlot$ClinicalSignificance[toPlot$ClinicalSignificance==""]<-NA
	clinvarMask <- ClinVarLabels$ClinicalSign %in% input$ClinVarInput
	
	# Plot
	figClinVar <- toPlot %>% plot_ly(x = ~physicalPos,
							   y = ~ClinicalSignificance,
							   color=~ClinicalSignificance,
							   colors=ClinVarLabels$colors[clinvarMask],
							   type="scatter",
							   mode="markers",
							   marker = list(
							   	size=10,
							   	line = list(color = '#404241', width = 1)),
							   text = ~paste0('<b>', rsid,'</b><br>',
							   			ClinicalSignificance, '<br> ', physicalPos),
							   hoverinfo = 'text') %>%
		layout(title="",
			  yaxis = list(
			  	title = '',
			  	type = 'category',
			  	autorange="reversed",
			  	categoryorder="array",
			  	categoryarray= ClinVarLabels$ClinicalSign[clinvarMask]),
			  xaxis = list(title = 'Position (bp)',
			  		   range = c(input$minPosInput, input$maxPosInput)),
			  showlegend = FALSE)
	
	figClinVar
    }) 
    output$clinvarPie <- renderPlotly({
    	
    	# Filter data
	toPlot <- dataSet() %>% dplyr::select(physicalPos, rsid, ClinicalSignificance, ClinSigSimple, PhenotypeList) %>%
		dplyr::filter(ClinicalSignificance %in% input$ClinVarInput, !is.na(rsid)) %>%
		dplyr::filter_at(vars(-c("physicalPos", "rsid")), any_vars(!is.na(.)))
	
	toPlot$ClinicalSignificance[toPlot$ClinicalSignificance==""]<-NA
	clinvarMask <- ClinVarLabels$ClinicalSign %in% input$ClinVarInput
    	numClin <- data.table(table(toPlot$ClinicalSignificance))
	names(numClin)<- c('ClinicalTrait','Counts')

	ClinVarpie <- plot_ly(numClin, labels = ~ClinicalTrait, values = ~Counts, type = 'pie',
                     textposition = 'inside',
                     textinfo = 'percent',
                     insidetextfont = list(color = '#FFFFFF'),
                     hoverinfo = 'text',
                     text = ~ClinicalTrait,
                     marker = list(colors = colors,
                                   line = list(color = '#FFFFFF', width = 1)),
                     #The 'pull' attribute can also be used to create space between the sectors
                     showlegend = FALSE)
      
      ClinVarpie <- ClinVarpie %>% layout(title = "ClinVar Consequences",
         xaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE),
         yaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE))

      ClinVarpie
    })
    
    output$gwasPlot <- renderPlotly({
      # Filter data
      toPlot <- dataSet() %>% dplyr::select( physicalPos, rsid, risk_allele, risk_allele_freq,
                                             trait, accession, context, NumGWASStudies) %>% 
        dplyr::filter_at(vars(-c("physicalPos", "rsid")), any_vars(!is.na(.))) %>% 
        dplyr::filter(between(NumGWASStudies, input$GWASInput[1], input$GWASInput[2]))
      
      # Plot
	    plot_GWAScat_scatter <- toPlot %>% plot_ly(x = ~physicalPos,
	                              y = ~NumGWASStudies,
	                              type="scatter",
	                              mode="markers",
	                              text = ~paste0('<b>', rsid,'</b>: ',
	                                             NumGWASStudies, ' study/es'),
	                              hoverinfo = 'text',
	                              colors = c('firebrick', 'darkblue'),
	                              showlegend = F,
	                              marker = list(size = 10,
	                                            color = 'rgba(255, 182, 193, .9)',
	                                            line = list(color = 'rgba(152, 0, 0, .8)',
	                                                        width = 2))) %>% 
	    	layout(
	    		xaxis = list(title = 'Position (bp)',
	    				   range = c(input$minPosInput, input$maxPosInput)))
	    
	    plot_GWAScat_scatter
	    })
    output$gwasPie <- renderPlotly({
      # Filter data
      toPlot <- dataSet() %>% dplyr::select( physicalPos, rsid, risk_allele, risk_allele_freq,
                                             trait, accession, context, NumGWASStudies) %>% 
        dplyr::filter_at(vars(-c("physicalPos", "rsid")), any_vars(!is.na(.))) %>% 
        dplyr::filter(between(NumGWASStudies, input$GWASInput[1], input$GWASInput[2]))
      
      numTraits <- data.table(table(unlist(strsplit(toPlot$trait, ";"))))
      
      fig <- plot_ly(numTraits, labels = ~V1, values = ~N, type = 'pie',
                     textposition = 'inside',
                     textinfo = 'percent',
                     insidetextfont = list(color = '#FFFFFF'),
                     hoverinfo = 'text',
                     text = ~V1,
                     marker = list(colors = colors,
                                   line = list(color = '#FFFFFF', width = 1)),
                     #The 'pull' attribute can also be used to create space between the sectors
                     showlegend = FALSE)
      
      fig <- fig %>% layout(title = "Traits appearing in GWAS catalog",
         xaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE),
         yaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE))

      fig
	})
    
    output$dgnPlot <- renderPlotly({
    	# Filter
    	toPlot <- dataSet() %>% dplyr::select(physicalPos, rsid, diseaseName, diseaseId, 
    										diseaseType, source, DSI, DPI, EI, NofPmids,NumDGNAsso) %>%
    		dplyr::filter_at(vars(-c("physicalPos", "rsid")), any_vars(!is.na(.))) %>% as.data.table
    	
    	toPlot = toPlot[DSI >= input$DSIInput[1] & DSI <= input$DSIInput[2] & 
    				 	EI >= input$EInput[1] & EI <= input$EInput[2] & 
    				 	NofPmids >= input$NofPmidsInput[1] & NofPmids <= input$NofPmidsInput[2]]
    	toPlot$NofPmids[is.na(toPlot$NofPmids)] <- 0
	toPlot$EI[is.na(toPlot$EI)] <- NULL

	# Plot
	fig <- toPlot %>% plot_ly(x = ~physicalPos,
						 y = ~DSI,
						 type = 'scatter', 
						 mode = 'markers',
						 marker = list(
						 	color=~EI,
						 	line = list(color = 'black', width = 2),
						 	size=10,
						 	colorbar = list(title = "Evidence Index"),
						 	colorscale='Viridis',
						 	showscale = TRUE),
						 text = ~paste0('<b>', rsid,'</b><br>',
						 			NumDGNAsso, 
						 			' associations(s)<br>',
						 			'<b>Type: </b>', gsub(";", ", ",diseaseType),
						 			'<br><b> Evidence Index:</b> ',ifelse(!is.na(EI), EI,"?"),
						 			'<br><br>', NofPmids, ' Pubmed Identifier(s)'),
						 hoverinfo = 'text') %>% 
		layout(title="",
			  yaxis = list(title = 'Disease Specificity Index',
			  		   range = c(0, 1.3)),
			  xaxis = list(title = 'Position (bp)',
				  		   range = c(input$minPosInput, input$maxPosInput)),
			  range = c(input$minPosInput, input$maxPosInput))
		fig
    })
    
    # Age plot
    output$agePlot <- renderPlotly({
    	
    	# Subset according to filters
	toPlot <- dataSet() %>% dplyr::select(physicalPos, rsid, ends_with(input$AgeModelInput)) %>% 
		dplyr::filter_at(vars(-c("physicalPos", "rsid")), any_vars(!is.na(.)))
	names(toPlot) <- gsub(paste0("_",input$AgeModelInput),"",names(toPlot))

	toPlot <- toPlot %>% dplyr::filter(QualScore>=input$qualScoreInput[1],
								QualScore<=input$qualScoreInput[2],
								AgeMode>=input$AgeModeInput[1],
								AgeMode<=input$AgeModeInput[2])
	# Adjust vars
	if(input$GenYears>0){
		toPlot <- toPlot %>% dplyr::mutate_at(vars(starts_with('Age')), ~. *input$GenYears)
		Yunits <- "years"
		GenLong <- paste0("Age\n(",input$GenYears," ", Yunits, "/generation)")
	}else{
		Yunits <- "generations"
		GenLong <- paste0("Age\n(generations)")
	}

	ClockLong <- GEVAclocks$plotTitle[GEVAclocks$modelClocks==input$AgeModelInput]

	# Plot the data
	figAge <- plot_ly(data = toPlot, x=~physicalPos)

	# If error bars marked
	if(input$errorBars){
		figAge <-figAge %>% add_trace(
			data=toPlot,
			y = ~AgeMode,
			type = 'scatter',
			mode = 'markers',
			name = ClockLong,
			size= ~QualScore,
			opacity= ~QualScore,
			color= ~QualScore,
			colors= colorByQuality(floor(min(toPlot$QualScore)*10),
							   ceiling(max(toPlot$QualScore)*10)),
			marker = list(line = list(color = '#26202b')),
			error_y =  list(type = "data",
						 symmetric = FALSE,
						 arrayminus = ~AgeCI95Lower,
						 array = ~AgeCI95Upper,
						 color = '#26202b',
						 size=0.05),
			showlegend = FALSE,
			text = ~paste0(
				'<b>', rsid,'</b><br>', physicalPos,
				"<br><b>",AgeMode," ", Yunits,"</b><br>(",
				round(AgeCI95Lower,2),"-", round(AgeCI95Upper,2),")"),
			hoverinfo = 'text')
		} else {
			# If not error bars
			figAge <-figAge %>% add_trace(
				data=toPlot,
				y = ~AgeMode,
				type = 'scatter',
				mode = 'markers',
				name = ClockLong,
				size= ~QualScore,
				opacity= ~QualScore,
				color= ~QualScore,
				colors= colorByQuality(floor(min(toPlot$QualScore)*10),
								   ceiling(max(toPlot$QualScore)*10)),
				marker = list(line = list(color = '#26202b')),
				showlegend = FALSE,
				text = ~paste0(
					'<b>', rsid,'</b><br>', physicalPos,
					"<br><b>",AgeMode," ", Yunits,"</b><br>(",
					round(AgeCI95Lower,2),"-", round(AgeCI95Upper,2),")"),
				hoverinfo = 'text')
			}
	figAge <-figAge %>%
		colorbar(title = "Quality Score",
			    thickness=15, len=0.25,tickmode ="auto",nticks=0,
			    bordercolor = '#26202b') %>%
		layout(xaxis = list(title = "Position (bp)", range = c(input$minPosInput, input$maxPosInput)),
			  yaxis = list(title = GenLong),
			  title=paste0(ClockLong))
	figAge
	
    })
    
    # Report Plot
    output$DescriptionRegionTitle <- renderText({
    	paste0("<h3>Region (GRCh37/hg19): <b>chr", chrN,":",
    		  input$minPosInput,"-",input$maxPosInput,'</b></h3>')
	})
    output$pophumanLink <- renderUI({
			actionButton("PHlinkbutton", 
				label = " PopHuman", 
				icon = icon("gitter"),
				onclick =paste0("window.open('https://pophuman.uab.cat/?loc=chr",
					chrN,"%3A",input$minPosInput,"..",input$maxPosInput,
					"&tracks=DNA%2Cgene_annotations%2Crecomb_Bherer2017_sexavg_10kb&highlight=', '_blank')"),
				width="95%",
				style="
				border: 1.8px solid #ffffff;
				color: #343a40;
				background-color: #f3c073;
				padding: 15px 32px;
				text-align: center;
				display: inline-block;
				font-size: 16px;
				margin: 10px 5px;
				cursor: pointer;
				border-radius: 5px;")
	})
    output$pophumanscanLink <- renderUI({

		linkToPHS <- paste0("window.open('https://pophumanscan.uab.cat/tables.php?coord=chr",
						chrN,':',input$minPosInput,'-',input$maxPosInput,"', '_blank')")

		actionButton("PHSlinkbutton", 
			label = " PopHumanScan", 
			icon = icon("fingerprint"),
			onclick = linkToPHS,
			width="95%",
			style="
			border: 1.8px solid #ffffff;
			color: #343a40;
			background-color: #f18686;
			padding: 15px 32px;
			text-align: center;
			display: inline-block;
			font-size: 16px;
			margin: 10px 5px;
			cursor: pointer;
			border-radius: 5px;")
	})
    
    set_report <- reactive({
    	# Generate data
    	filtered <- dataSet() %>% dplyr::select(physicalPos, rsid,standarizediHS, standarizednSL,
    									iSAFE, effect, impact,'RDBranking'=ranking, ClinicalSignificance,
    									DSI, EI, NumGWASStudies,
    									paste0(c('AgeMode_','QualScore_','AgeCI95Lower_', 'AgeCI95Upper_'),
    										  input$Overview_Clock))
    	# Change columns to ease operate with them
	names(filtered) = gsub(paste0("_", input$Overview_Clock),"", names(filtered))

	# Filter according the side gadgets
	filtered = filtered[AgeMode > input$Overview_AgeMode[1] & AgeMode < input$Overview_AgeMode[2]]
	filtered = filtered[QualScore > input$Overview_qualScore[1] & QualScore < input$Overview_qualScore[2]]

	# Modify dataSet() to obtain Plot
	expandDF <- SplitDf(filtered, c1=c("effect","impact"), c2=c("effect"),s1=",",s2="&")
	expandDF$rank <- 0
	for(i in unique(na.omit(expandDF$effect))){
		expandDF[effect == i]$rank = VEPinfo[inLabels == i]$rank
		}
	
	expandDF$rank[expandDF$rank==0]<-NA
	toPlot <- expandDF[expandDF[,.I[which.min(rank)], by=list(physicalPos, rsid)]$V1]
	toPlot$iSAFE[is.na(toPlot$iSAFE)]<- 0
	toPlot$standarizediHS[is.na(toPlot$standarizediHS)]<- 0
	toPlot$standarizednSL[is.na(toPlot$standarizednSL)]<- 0
	
	toPlot
    })
    
    set_prior <- reactive({
    	# Sort and change columns names
    	toFilter <- set_report() %>%
    		dplyr::arrange(
    			desc(iSAFE),
    			desc(abs(standarizediHS)),
    			desc(abs(standarizednSL)),
    			desc(NumGWASStudies),
    			rank,
    			RDBranking) %>%
    		dplyr::select(
    			'Position'=physicalPos,
    			rsid,
    			iSAFE,
    			'iHS'=standarizediHS,
    			'nSL'=standarizednSL,
    			'GWAS assoc.'=NumGWASStudies,
    			'Variant effect'=effect,
    			'Regulome Score'=RDBranking,
    			'Generations'=AgeMode
			)
    	# Subset top 20
    	finalSubset <-toFilter[1:20,]
    	finalSubset$'Variant effect' <- str_to_title(gsub("_"," ",finalSubset$'Variant effect'))
    	
    	finalSubset
	})
    
    output$reportPlot <- renderPlotly({
		# Add data to plot
		toPlot = set_report()

		# Set thresholds
		tresh_ihs   = unname(quantile(abs(toPlot$standarizediHS),1-input$Overview_pvalue_ihs, na.rm=TRUE))
		tresh_nsl   = unname(quantile(abs(toPlot$standarizednSL),1-input$Overview_pvalue_nsl, na.rm=TRUE))
		tresh_isafe = unname(quantile(toPlot$iSAFE,1-input$Overview_pvalue_isafe, na.rm=TRUE))

		# PLOT!
		fig <- toPlot %>%
			plot_ly(
				x = ~physicalPos,
				y = ~iSAFE,
				type = 'scatter',
				mode = 'markers',
				color = ~str_to_title(gsub("_"," ", effect)),
				colors='Paired',
				size = ~abs(standarizediHS)+abs(standarizednSL),
				sizes = c(7, 25),
				marker = list(opacity = 0.75, sizemode = 'diameter', 
								line = list(width = 1.25, color = '#000000')),
				text = ~paste0('<b>', rsid,'</b> &#8594; ', physicalPos,'<br><br>',
						ifelse(iSAFE!=0, paste0("<b>iSAFE:</b> ",round(iSAFE,3),'<br>'),""),
						ifelse(iSAFE>=tresh_isafe, paste0(" <b>(!) top ",input$Overview_pvalue_isafe*100,"%</b><br>"),""),
						ifelse(standarizediHS!=0, paste0("<br><b>iHS:</b> ",round(standarizediHS,3),'<br>'),""),
						ifelse(standarizediHS>=tresh_ihs, paste0(" <b>(!) top ",input$Overview_pvalue_ihs*100,"%</b><br>"),""),
						ifelse(standarizednSL!=0, paste0("<br><b>nSL:</b> ",round(standarizednSL,3),'<br>'),""),
						ifelse(standarizednSL>=tresh_nsl, paste0(" <b>(!) top ",input$Overview_pvalue_nsl*100,"%</b><br>"),""),
						'<br><b>',str_to_title(gsub("_"," ", effect)),'</b> (',tolower(impact), ')<br>',
						ifelse(!is.na(ClinicalSignificance), paste0(ClinicalSignificance,"<br>"),""),
						ifelse(!is.na(NumGWASStudies), paste0("GWAS hits:",NumGWASStudies,"<br>"),""),
						ifelse(!is.na(DSI), paste0("Disease sp. index:",DSI,"<br>"),""),
						ifelse(!is.na(AgeMode), paste0("<b>Age: ", AgeMode, " gen.</b> [",
							AgeCI95Lower,"-",AgeCI95Upper,"] (QS=", QualScore,")<br>"),""),"<br>"
						),
			hoverinfo = 'text')
			
		# Check if the table is highlighted
		s = input$tablePrior_rows_selected

		# If any, add annotations
		if(length(s)){
			m <- set_prior()[s,]
			 fig <- fig %>% layout(
			 	annotations = list(
			 		x = m$Position,
			 		y = m$iSAFE,
			 		text = paste0("<b>",s,":</b> ",m$rsid),
			 		text = "Center Anchor",
			 		showarrow = T,
			 		# Styling annotations' text:
			 		font = list(size = 14)
			 		)
			 )
		 }

		fig <- fig %>%
			layout(legend = list(orientation = 'h', y = -0.3,
							 itemsizing='constant'),
				  xaxis = list(title = 'Position (bp)',
				  		   range = c(input$minPosInput, input$maxPosInput)),
				  yaxis = list(title = 'iSAFE'))

		if(max(toPlot$iSAFE, na.rm = TRUE) > tresh_isafe){
			fig <- fig %>%
				layout(
					shapes = list(
						type = "line",
						x0 = 0,
						x1 = 1,
						xref = "paper",
						y0 = tresh_isafe,
						y1 = tresh_isafe,
						line = list(color = "orange"),
						width = 4,
						dash = 'dot',
						opacity = 0.4
						)
				)
			}
		fig
	})
    
    haplo_plots <- reactive({

    	#get selected rsid if non selected get first
		idx = input$tablePrior_rows_selected
		if(is.null(idx)){
			idx = 1
		}

		# load rehh file
		hh = data2haplohh(list.files(pattern = '\\.vcf\\.gz')[1],polarize_vcf=F)

		# rehh plot's estimations
		ehh       = calc_ehh(hh,mrk=set_prior()[idx,2])
		furcation = calc_furcation(hh,mrk=set_prior()[idx,2])
		haplen    = calc_haplen(furcation)

		# return objects to plot
		list(ehh,furcation,haplen)
	})

	output$haplo1 	<- renderPlot({
		plot(haplo_plots()[[1]])
	})
	output$haplo2 <- renderPlot({
		plot(haplo_plots()[[2]])
	})
	output$haplo3 <- renderPlot({
		plot(haplo_plots()[[3]])
	})
    
    
    # Subset prioretized regions
	set_infoBoxes<-reactive({
		# Sort and change columns names
		# Generate data
		filtered <- dataSet() %>% 
			dplyr::select(physicalPos, rsid,standarizediHS, standarizednSL,
    									iSAFE, NumGWASStudies, effect, 
					    impact,'RDBranking'=ranking, ClinicalSignificance, diseaseId,
					    NofPmids,paste0(c('AgeMode_','QualScore_','AgeCI95Lower_', 'AgeCI95Upper_'), input$Overview_Clock))
		
		# Change columns to ease operate with them
		names(filtered) = gsub(paste0("_", input$Overview_Clock),"", names(filtered))

		# Filter according the side gadgets
		filtered = filtered[AgeMode > input$Overview_AgeMode[1] & AgeMode < input$Overview_AgeMode[2]]
		filtered = filtered[QualScore > input$Overview_qualScore[1] & QualScore < input$Overview_qualScore[2]]
			
		# Modify dataset to obtain Plot
		expandDF <- SplitDf(filtered, c1=c("effect","impact"), c2=c("effect"),s1=",",s2="&")
		expandDF$rankSnpEFF <- 0
		for(i in unique(na.omit(expandDF$effect))){expandDF[effect == i]$rankSnpEFF = VEPinfo[inLabels == i]$rank}
		expandDF$rankSnpEFF[expandDF$rankSnpEFF==0]<-NA	
		toPlot <- expandDF[expandDF[,.I[which.min(rankSnpEFF)], by=list(physicalPos, rsid)]$V1]
		
		toPlot$isafe[is.na(toPlot$iSAFE)]<-0
		toPlot$standarizediHS[is.na(toPlot$standarizediHS)]<-0
		toPlot$standarizednSL[is.na(toPlot$standarizednSL)]<-0
		toPlot$ihsABS <- abs(toPlot$standarizediHS)
		toPlot$nslABS <- abs(toPlot$standarizednSL)
		toPlot$ClinVarRank <- match(toPlot$ClinicalSignificance, ClinVarLabels$ClinicalSign, nomatch=NA)
		
		toShow <- toPlot %>% dplyr::select(physicalPos, rsid, 'isafe'=iSAFE, 
									'ihs'=standarizediHS, ihsABS, 'nsl'=standarizednSL, nslABS, 
									NumGWASStudies, rankSnpEFF, effect, impact, RDBranking,
									ClinicalSignificance, ClinVarRank, diseaseId, NofPmids)

		toShow
	})
    
    
    	############ INFO BOX WITH SUMMARY
	output$info_isafe <- renderInfoBox({
		tresh_isafe = unname(quantile(set_infoBoxes()$iSAFE,1-input$Overview_pvalue_isafe, na.rm=TRUE))
		info <- as.vector(unlist(getInfoBoxData('isafe', set_infoBoxes())))
		if(is.null(info)){
			infoBox(
				title    = HTML('Favored Mutation: <b><span style="text-transform:lowercase">i</span><span style="text-transform:uppercase">SAFE</span></b>'),
				subtitle="for iSAFE in this region",
				value="No data",
				icon     = icon("asterisk"),
				color    = "olive"
				)
		} else {
			if(as.numeric(info[3])==1){
					infoBox(
						title    = HTML('Favored Mutation: <b><span style="text-transform:lowercase">i</span><span style="text-transform:uppercase">SAFE</span></b>'),
						value=HTML(paste0("<a href='https://www.ncbi.nlm.nih.gov/snp/",info[1],"' target='_blank'>",info[1],"</a>")),
						subtitle=paste0("reports the most extreme iSAFE value: ", round(as.numeric(info[2]),4),
							ifelse(as.numeric(info[2])>tresh_isafe,
								" and belongs to the top 0.01%","")
							),
						icon     = icon("asterisk"),
						color    = "olive")
			} else{
				infoBox(
					title    = HTML('Favored Mutation: <b><span style="text-transform:lowercase">i</span><span style="text-transform:uppercase">SAFE</span></b>'),
					value = HTML(paste0("<a href='https://www.ncbi.nlm.nih.gov/snp/",info[1],"' target='_blank'>",info[1],"</a>")),
					subtitle=paste0("... and ", as.numeric(info[3]) - 1," more report the most extreme iSAFE value: ", round(as.numeric(info[2]),4),
						ifelse(as.numeric(info[2])>tresh_isafe,
							" and belong to the top 0.01%","")
						),
					icon     = icon("asterisk"),
					color    = "olive")
				}
		}
	})
	output$info_iHS <- renderInfoBox({
		tresh_ihs   = unname(quantile(abs(set_infoBoxes()$ihs),1-input$Overview_pvalue_ihs, na.rm=TRUE))
		info <- as.vector(unlist(getInfoBoxData('ihs', set_infoBoxes())))
		if(is.null(info)){
			infoBox(
				title= HTML('Selection: <b><span style="text-transform:lowercase">i</span><span style="text-transform:uppercase">HS</span></b>'),
				subtitle="for iHS in this region",
				value="No data",
				icon = icon("stream"),
				color = "yellow"
				)
		} else {
			if(as.numeric(info[3])==1){
				infoBox(
					title    = HTML('Selection: <b><span style="text-transform:lowercase">i</span><span style="text-transform:uppercase">HS</span></b>'),
					value=HTML(paste0("<a href='https://www.ncbi.nlm.nih.gov/snp/",info[1],"' target='_blank'>",info[1],"</a>")),
					subtitle=paste0("reports the most extreme iHS value: ", round(as.numeric(info[2]),2),
						ifelse(as.numeric(info[2])>tresh_ihs,
							" and belongs to the top 0.5%","")
						),
					icon = icon("stream"),
					color = "yellow")
				} else{
					infoBox(
						title    = HTML('Selection: <b><span style="text-transform:lowercase">i</span><span style="text-transform:uppercase">HS</span></b>'),
						value=HTML(paste0("<a href='https://www.ncbi.nlm.nih.gov/snp/",info[1],"' target='_blank'>",info[1],"</a>")),
						subtitle=paste0("... and ",as.numeric(info[3]) - 1," more report the most extreme iHS value: ", round(as.numeric(info[2]),2),
							ifelse(as.numeric(info[2])>tresh_ihs,
								" and belong to the top 0.5%","")
							),
						icon = icon("stream"),
						color = "yellow")
				}
			}
	})
	output$info_nsl <- renderInfoBox({

		tresh_nsl   = unname(quantile(abs(set_infoBoxes()$nsl),1-input$Overview_pvalue_nsl, na.rm=TRUE))
		info <- as.vector(unlist(getInfoBoxData('nsl', set_infoBoxes())))

		if(is.null(info)){
			infoBox(
				title    = HTML('Selection: <b><span style="text-transform:lowercase">n</span><span style="text-transform:uppercase">SL</span></b>'),
				subtitle="for nSL in this region",
				value="No data",
				icon = icon("stream"),
				color = "red"
				)
			} else {
				if(as.numeric(info[3])==1){
					infoBox(
						title    = HTML('Selection: <b><span style="text-transform:lowercase">n</span><span style="text-transform:uppercase">SL</span></b>'),
						value=HTML(paste0("<a href='https://www.ncbi.nlm.nih.gov/snp/",info[1],"' target='_blank'>",info[1],"</a>")),
						subtitle=paste0("reports the most extreme nSL value: ", round(as.numeric(info[2]),2),
							ifelse(as.numeric(info[2])>tresh_nsl,
								" and belongs to the top 0.5%","")
							),
						icon = icon("stream"),
						color = "red"
						)
					} else{
						infoBox(
							title    = HTML('Selection: <b><span style="text-transform:lowercase">n</span><span style="text-transform:uppercase">SL</span></b>'),
							value=HTML(paste0("<a href='https://www.ncbi.nlm.nih.gov/snp/",info[1],"' target='_blank'>",info[1],"</a>")),
							subtitle=paste0("... and ",as.numeric(info[3]) - 1," more report the most extreme nSL value: ", round(as.numeric(info[2]),2),
								ifelse(as.numeric(info[2])>tresh_nsl,
									" and belong to the top 0.5%","")
								),
							icon = icon("stream"),
							color = "red"
							)
					}
				}
	})
	output$info_snpeff <- renderInfoBox({
		info <- as.vector(unlist(getInfoBoxData('snpeff', set_infoBoxes())))
		if(is.null(info)){
			infoBox(
				title    = HTML('Functional: <b>S<span style="text-transform:lowercase">np</span><span style="text-transform:uppercase">EFF</span></b>'),
				subtitle="for SnpEFF in this region",
				value="No data",
				icon = icon("layer-group"),
				color = "purple")
		} else {
			if(as.numeric(info[3])==1){
				infoBox(
					title    = HTML('Functional: <b>S<span style="text-transform:lowercase">np</span><span style="text-transform:uppercase">EFF</span></b>'),
					value=HTML(paste0("<a href='https://www.ncbi.nlm.nih.gov/snp/",info[1],"' target='_blank'>",info[1],"</a>")),
					subtitle=paste0("reports the most strong impact: ", str_to_title(gsub("_"," ",info[2]))),
					icon = icon("layer-group"),
					color = "purple")
				} else{
					infoBox(
						title    = HTML('Functional: <b>S<span style="text-transform:lowercase">np</span><span style="text-transform:uppercase">EFF</span></b>'),
						value=HTML(paste0("<a href='https://www.ncbi.nlm.nih.gov/snp/",info[1],"' target='_blank'>",info[1],"</a>")),
						subtitle=paste0("... and ",as.numeric(info[3]) - 1," other report the most strong impact: ", str_to_title(gsub("_"," ",info[2]))),
						icon = icon("layer-group"),
						color = "purple")
			}
		}
	})
	output$info_ReguDB <- renderInfoBox({

		info <- as.vector(unlist(getInfoBoxData('rgd', set_infoBoxes())))

		if(is.null(info)){
			infoBox(
				title    = HTML('Functional: <b>R<span style="text-transform:lowercase">egulome</span><span style="text-transform:uppercase">DB</span></b>'),
				subtitle="for RegulomeDB in this region",
				value="No data",
				icon("layer-group"),
				color = "navy")
			} else {
				if(as.numeric(info[3])==1){
					infoBox(
						title    = HTML('Functional: <b>R<span style="text-transform:lowercase">egulome</span><span style="text-transform:uppercase">DB</span></b>'),
						value=HTML(paste0("<a href='https://regulomedb.org/regulome-summary/?regions=",info[1],"' target='_blank'>",info[1],"</a>")),
						subtitle=paste0("reports the highest RegulomeDB score: ", info[2]),
						icon("layer-group"),
						color = "navy")
					} else{
						infoBox(
							title    = HTML('Functional: <b>R<span style="text-transform:lowercase">egulome</span><span style="text-transform:uppercase">DB</span></b>'),
							# value=HTML(paste0("<a href='https://www.ncbi.nlm.nih.gov/snp/",info[1],")),
							value=HTML(paste0("<a href='https://regulomedb.org/regulome-summary/?regions=",info[4],"&genome=GRCh37&maf=0.01' target='_blank'>",info[1],"</a>")),
							subtitle=paste0("... and ",as.numeric(info[3]) - 1," other report the highest RegulomeDB score:  ", info[2]),
						icon("layer-group"),
						color = "navy")
						}
					}
	})
	output$info_ClinVar <- renderInfoBox({

		info <- as.vector(unlist(getInfoBoxData('clinvar', set_infoBoxes())))

		if(is.null(info)){
			infoBox(
				title    = HTML('Functional: <b>C<span style="text-transform:lowercase">lin</span>V<span style="text-transform:lowercase">ar</span></b>'),
				subtitle="for ClinVar in this region",
				value="No data",
				icon = icon("layer-group"),
				color = "teal")
			} else {
				if(as.numeric(info[3])==1){
					infoBox(
						title    = HTML('Functional: <b>C<span style="text-transform:lowercase">lin</span>V<span style="text-transform:lowercase">ar</span></b>'),
						value=HTML(paste0("<a href='https://www.ncbi.nlm.nih.gov/snp/",info[1],"' target='_blank'>",info[1],"</a>")),
						subtitle=paste0("reports the interpretation: ", str_to_title(gsub("_"," ",info[2]))),
						icon = icon("layer-group"),
						color = "teal")
					} else{
						infoBox(
							title    = HTML('Functional: <b>C<span style="text-transform:lowercase">lin</span>V<span style="text-transform:lowercase">ar</span></b>'),
							value=HTML(paste0("<a href='https://www.ncbi.nlm.nih.gov/snp/",info[1],"' target='_blank'>",info[1],"</a>")),
							subtitle=paste0("... and ",as.numeric(info[3]) - 1," other report the interpretation: ", str_to_title(gsub("_"," ",info[2]))),
							icon = icon("layer-group"),
							color = "teal")
					}
				}
	})
	output$info_DGN <- renderInfoBox({

		info <- as.vector(unlist(getInfoBoxData('dgn', set_infoBoxes())))

		if(is.null(info)){
			infoBox(
				title    = HTML('Functional: <b>D<span style="text-transform:lowercase">is</span>G<span style="text-transform:lowercase">e</span>NET</b>'),
				subtitle="for DisGeNET in this region",
				value="No data",
				icon("layer-group"),
				color = "green"
				)
			} else {
				if(as.numeric(info[3])==1){
					infoBox(
						title    = HTML('Functional: <b>D<span style="text-transform:lowercase">is</span>G<span style="text-transform:lowercase">e</span>NET</b>'),
						value=HTML(paste0("<a href='https://www.disgenet.org/browser/2/1/0/",info[1],"/' target='_blank'>",info[1],"</a>")),
						subtitle=paste0("reports the highest number of associated PubmedIDs: ", info[2]),
						icon("layer-group"),
						color = "green")
					} else{
						infoBox(
							title    = HTML('Functional: <b>D<span style="text-transform:lowercase">is</span>G<span style="text-transform:lowercase">e</span>NET</b>'),
							value=HTML(paste0("<a href='https://www.disgenet.org/browser/2/1/0/",info[1],"/' target='_blank'>",info[1],"</a>")),
							subtitle=paste0("... and ",as.numeric(info[3]) - 1," other report the highest number of associated PubmedIDs: ", info[2]),
							icon("layer-group"),
							color = "green")
					}
				}
	})
	output$info_GWAScat <- renderInfoBox({

		info <- as.vector(unlist(getInfoBoxData('gwas', set_infoBoxes())))

		if(is.null(info)){
			infoBox(
				title    = HTML('Functional: <b>GWAS C<span style="text-transform:lowercase">atalog</span></b>'),
				subtitle="for GWAS Catalog in this region",
				value="No data",
				icon("layer-group"),
				color = "maroon")
			} else {
				if(as.numeric(info[3])==1){
					infoBox(
						title    = HTML('Functional: <b>GWAS C<span style="text-transform:lowercase">atalog</span></b>'),
						value=HTML(paste0("<a href='https://www.ebi.ac.uk/gwas/variants/",info[1],"' target='_blank'>",info[1],"</a>")),
						subtitle=paste0("reports the highest number of hits: ", info[2]),
						icon("layer-group"),
						color = "maroon")
					} else{
						infoBox(
							title    = HTML('Functional: <b>GWAS C<span style="text-transform:lowercase">atalog</span></b>'),
							value=HTML(paste0("<a href='https://www.ebi.ac.uk/gwas/variants/",info[1],"' target='_blank'>",info[1],"</a>")),
							subtitle=paste0("... and ",as.numeric(info[3]) - 1," other report the highest number of hits: ", info[2]),
							icon("layer-group"),
							color = "maroon")
					}
				}
	})


	# Table with top 20 variants
	output$tablePrior <- DT::renderDataTable(
		set_prior(),
		options = list(pageLength = 20,
					scrollX=TRUE),
		selection ='single',
		server = FALSE
	)
	
     # Show data Table
    output$table_all <- renderDT(
      dataSet(),
      selection = 'none', 
      options = list(pageLength = 50,
                     scrollX=TRUE))
  },
  options = list(width = "120%", height = 2200)
)
```
